networkit/cpp/io/NetworkitBinaryWriter.cpp

```diff
@@ -0,0 +1,211 @@
+
+	std::ofstream outfile(path, std::ios::binary);
+	Aux::enforceOpened(outfile);
+
+	bool directed = G.isDirected();
+	uint64_t format = 0;
+	unsigned weight = NONE;
+	Header header = {};
+
+	auto calculateBaseOffset = [&] () {
+		return sizeof(Header);
+	};
+
+	auto calculateAdjOffset = [&] () {
+		size_t off = nodes * sizeof(uint64_t); //prefSum
+		off += chunks * sizeof(uint64_t); //first vertex
+		return sizeof(Header) + off;
+	};
+
+	auto calculateWeightsOffset = [&] () {
+		size_t off = G.numberOfEdges() * sizeof(uint8_t); //adj arrays
+		off += chunks * sizeof(uint64_t);		  // adj offsets
+		return sizeof(Header) + off + calculateAdjOffset();
+	};
+
+	auto setFeatures = [&] () {
+		header.features |= (G.isDirected() & DIR_MASK);
+		header.features |= ((weight << WGHT_SHIFT) & WGHT_MASK);
+	};
+
+	auto setHeader = [&] () {
+		strncpy(header.magic,"nkbg000",8);
+		header.checksum = 0;
+		setFeatures();
+		header.nodes = nodes;
+		header.chunks = chunks;
+		header.offsetBaseData = calculateBaseOffset();
+    	header.offsetAdjLists = calculateAdjOffset();
+    	header.offsetWeights = calculateWeightsOffset();
+	};
+
+	auto writeHeader = [&] () {
+
+		outfile.write(header.magic, 8);
+		outfile.write(reinterpret_cast<char*>(&header.checksum), sizeof(uint64_t));
+		outfile.write(reinterpret_cast<char*>(&header.features), sizeof(uint64_t));
+		outfile.write(reinterpret_cast<char*>(&header.nodes), sizeof(uint64_t));
+		outfile.write(reinterpret_cast<char*>(&header.chunks), sizeof(uint64_t));
+		outfile.write(reinterpret_cast<char*>(&header.offsetBaseData), sizeof(uint64_t));
+		outfile.write(reinterpret_cast<char*>(&header.offsetAdjLists), sizeof(uint64_t));
+		outfile.write(reinterpret_cast<char*>(&header.offsetWeights), sizeof(uint64_t));
+	};
+
+	if(!G.isEmpty()) {
+		count prefSum = 0;
+		nodes = G.numberOfNodes();
+		if (nodes < chunks) {
+			chunks = nodes;
+			INFO("reducing chunks to ", chunks, " chunks");
+		}
+		nodesPerChunk = splitNodes(nodes, chunks);
+
+		size_t node_size = 0;
+		count sum = 0;
+		std::vector<count> nodePrefSum;
+		if(!G.isDirected()) {
+			G.forNodes([&](node u) {
+				std::vector<uint64_t> nbr;
+				uint8_t tmp [10];
+				node_size = NetworkitBinaryWriter::encode(u,tmp);
+				G.forNeighborsOf(u,[&](node v){
+					nbr.push_back(v);
+				});
+				sum += nbr.size();
+				nodePrefSum.push_back(sum);
+			});
+		} else {
+			G.forNodes([&](node u) {
+				std::vector<uint64_t> nbr;
+				G.forNeighborsOf(u,[&](node v){
+					nbr.push_back(v);
+				});
+				sum += nbr.size();
+				nodePrefSum.push_back(sum);
+			});
+		}
+
+		/* compute encoded size of arrays and store in vector */
+		size_t prev = 0;
+		std::vector<size_t> encArrSize; //  pref sum of size encoded adj arrays
+		G.forNodes([&](node u) {
+			size_t adj_size = 0;
+			G.forNeighborsOf(u,[&](node v) {
+				uint8_t tmp [10];
+				adj_size += encode(v, tmp);
+			});
+			adj_size += prev;
+			encArrSize.push_back(adj_size);
+			prev = adj_size;
+		});
```

Comment:
This could also be computed in the loop above, avoiding an iteration over the whole graph.

Futher: why is this computing the offset for every node? The chunk offsets are completely independent of the edges, therefore the chunk offsets could also be determined before, then only the sizes for each chunk would need to be stored.